index.js

```diff
@@ -0,0 +1,119 @@
+'use strict';
+/* eslint-disable no-underscore-dangle */
+const Joi = require('joi');
+
+class ScmBase {
+    /**
+     * Constructor for Scm
+     * @method constructor
+     * @param  {Object}    config Configuration
+     * @return {ScmBase}
+     */
+    constructor(config) {
+        this.config = config;
+    }
+
+    /**
+     * Reload configuration
+     * @method configure
+     * @param  {Object}     config      Configuration
+     */
+    configure(config) {
+        this.config = config;
+    }
+
+    /**
+     * Get a users permissions on a repository
+     * @method getPermissions
+     * @param  {Object}   config            Configuration
+     * @param  {String}   config.repo       The repo to get permissions on
+     * @param  {String}   config.user       The user to get pemissions for
+     * @param  {String}   config.token      The token used to authenticate to the SCM
+     * @return {Promise}
+     */
+    getPermissions(config) {
+        const result = Joi.validate(config, {
+            repo: Joi.string().required(),
+            user: Joi.string().required(),
+            token: Joi.string().required()
+        });
+
+        if (result.error) {
+            return new Promise((resolve, reject) => {
+                process.nextTick(() => reject(result.error));
+            });
+        }
+
+        return this._getPermissions(config);
+    }
+
+    _getPermissions() {
+        return new Promise((resolve, reject) => {
+            process.nextTick(() => reject('not implemented'));
+        });
+    }
+
+    /**
+     * Get a commit sha for a specific repo#branch
+     * @method getCommitSha
+     * @param  {Object}   config            Configuration
+     * @param  {String}   config.repo       The repo to get commit sha of
+     * @param  {String}   config.token      The token used to authenticate to the SCM
+     * @return {Promise}
+     */
+    getCommitSha(config) {
+        const result = Joi.validate(config, {
+            repo: Joi.string().required(),
+            token: Joi.string().required()
+        });
+
+        if (result.error) {
+            return new Promise((resolve, reject) => {
+                process.nextTick(() => reject(result.error));
+            });
+        }
+
+        return this._getCommitSha(config);
+    }
+
+    _getCommitSha() {
+        return new Promise((resolve, reject) => {
+            process.nextTick(() => reject('not implemented'));
+        });
+    }
+
+    /**
+     * Update the commit status for a given repo and sha
+     * @method get
+     * @param  {Object}   config            Configuration
```

Comment:
You get the user/repo from parsing `config.repo`

As for the user, I was thinking that we should be passing in tokens to the scm-base, and not a user object that you then need to unseal the token from